server/src/main/java/org/apache/druid/server/coordinator/loading/LoadingRateTracker.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.server.coordinator.loading;
+
+import com.google.common.collect.EvictingQueue;
+
+import javax.annotation.concurrent.ThreadSafe;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Tracks the current segment loading rate for a single server.
+ * <p>
+ * The loading rate is computed as a moving average of the last
+ * {@link #MOVING_AVERAGE_WINDOW_SIZE} progress updates (or more if any of the
+ * updates was smaller than {@link #MIN_ENTRY_SIZE_BYTES}).
+ */
+@ThreadSafe
+public class LoadingRateTracker
+{
+  public static final int MOVING_AVERAGE_WINDOW_SIZE = 10;
+  public static final long MIN_ENTRY_SIZE_BYTES = 1_000_000_000;
+
+  private final EvictingQueue<Entry> window = EvictingQueue.create(MOVING_AVERAGE_WINDOW_SIZE);
+  private final AtomicReference<Entry> windowTotal = new AtomicReference<>(new Entry());
+  private Entry currentTail;
+
+  public synchronized void updateProgress(long bytes, long millisElapsed)
+  {
+    if (bytes >= 0 && millisElapsed > 0) {
```

Comment:
Should this be `bytes > 0`?

Reply:
Yeah, I think that would be fair.

Although, I am not sure about tombstone segments. They would have size zero but it would still take non-zero time to load time. I guess in that case it would make sense to account for the time in the overall window average.

Let me know what you think.